Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp

```diff
@@ -678,22 +678,66 @@ static inline bool canCacheMeasuredWidthOnInlineTextItem(const InlineTextBox& in
     return !inlineTextBox.hasPositionDependentContentWidth();
 }
 
+class TextAutospaceAdjuster {
+public:
+    bool trackBoxStart(InlineItem& candidate)
+    {
+        if (candidate.isInlineBoxStart()) {
+            m_unhandledBoxStart = &candidate;
+            return true;
+        }
+        return false;
+    }
+
+    void adjustStartboxOffset(const InlineTextBox& inlineTextBox, unsigned start)
+    {
+        if (!m_unhandledBoxStart)
+            return;
+        auto textAutoSpace = inlineTextBox.parent().style().textAutospace();
+        if (textAutoSpace.hasIdeographAlpha() && textAutoSpace.shouldApplySpacing(m_spacingState.lastCharacterClassFromPreviousRun, TextSpacing::characterClass(inlineTextBox.content().characterAt(start)))) {
+            m_unhandledBoxStart->setOffset(TextAutospace::textAutospaceSize(inlineTextBox.parent().style().fontCascade().primaryFont()));
+            // https://www.w3.org/TR/css-text-4/#text-autospace-property
+            // "At element boundaries, the amount of extra spacing introduced between characters is determined by and rendered within the innermost element that contains the boundary."
+            // Reset startbox tracking and SpacingState here. We don't measure the width of the first character in the inlineTextBox considering its adjacency to the last character in the previous inlineTextBox, as this extra spacing is accounted for by the startBox offset above, as dictated by specification.
+            m_unhandledBoxStart = nullptr;
+            m_spacingState.lastCharacterClassFromPreviousRun = TextSpacing::CharacterClass::Undefined;
+        }
+    }
+
+    const TextSpacing::SpacingState& spacingState() const { return m_spacingState; }
+    void updateSpacingState(const InlineTextBox& inlineTextBox, unsigned start, unsigned length)
+    {
+        if (inlineTextBox.style().textAutospace().isNoAutospace())
+            m_spacingState.lastCharacterClassFromPreviousRun = TextSpacing::CharacterClass::Undefined;
+        else
+            m_spacingState.lastCharacterClassFromPreviousRun = TextSpacing::characterClass(inlineTextBox.content().characterAt(start + length - 1));
+    }
+
+private:
+    TextSpacing::SpacingState m_spacingState { TextSpacing::CharacterClass::Undefined };
+    InlineItem* m_unhandledBoxStart { nullptr };
+};
+
 void InlineItemsBuilder::computeInlineTextItemWidths(InlineItemList& inlineItemList)
 {
-    TextSpacing::SpacingState spacingState;
+    TextAutospaceAdjuster textAutospaceAdjuster;
```

Comment:
is this going to be a perf concern for content with no space adjustment at all?

Reply:
I haven't measured a perceived performance for content with `text-autospace: no`. I'm measuring again for the most recent iteration of this patch.